### PR TITLE
Upload new data files

### LINKED ISSUE #738: Add/update data tables from TSV
Users need to create new tables in Saturn representing their data. The most common way to do this is by uploading a tabular file exported from a program like Excel.

In the data tab, there should be an "Add Data" affordance. Ultimately, users will be able to add either tables or files here, but for now, we'll just handle the case of tables.

There are two things a user will need to do:
1) Create a new data table
2) Add data to an existing table.

In (1), the user would upload their flat file. Currently, TSV is the only format supported by the backend. (In the future we'll also likely want CSV). The table name (entity type) should default to the name of the file. It would be nice if the user could also optionally specify a different name.

In (2), the user would select an existing table, and upload a flat file. Columns and rows not present in the existing table would be appended. Columns and rows already present in the existing table would be overwritten with the new values from the file.
https://software.broadinstitute.org/firecloud/documentation/article?id=10892

The user MUST be able to specify custom table names (entity types). We will not restrict users to uploading tables corresponding to participants/samples/pairs.

In either case (1) or (2), we place some requirements on the format of the file. We should either inform the user of these requirements, allow them to download a blank template, or link to documentation about format restrictions. (Or all of the above).

For example, we require the first row of the file be the intended column headers. We also need to highlight a primary key column, which contains the unique identifiers for each row. This column cannot contain duplicate values. We require users to call out this column by including "entity:" as a prefix on the column name. We may also require that it end in "_id"? There are also custom formatting semantics for specifying relationships between fields of different tables.
See the 'sample' example here:
https://software.broadinstitute.org/firecloud/documentation/article?id=9771
@ahaessly Is there anything else notable about the file format here?

### PR DESCRIPTION
**Pull request by zarsky-broad**
closes #738

### CODE REVIEW

**Comment by zarsky-broad** [comment]
👍'ed by @lstrano-broad

File: src/libs/data-utils.js [hunk 1]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+
+    const inputLabel = text => div({ style: { fontSize: 16, marginBottom: '0.3rem' } }, [text])
+
+    return h(Dropzone, {
+      accept: '.tsv',
+      disableClick: true,
+      style: { flexGrow: 1 },
+      activeStyle: { backgroundColor: colors.blue[3], cursor: 'copy' },
+      ref: this.uploader,
+      onDropRejected: () => this.setState({ file: undefined, isInvalidFile: true, isInvalidTSV: false }),
```
Comment (panentheos) [inline src/libs/data-utils.js]: nit: since these error states are mutually exclusive, they could be modeled with a single nullable 'enum' string.

**Review by panentheos — APPROVED** [review]
Looks good.

File: src/libs/data-utils.js [hunk 2]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+    }
+  }
+
+  render() {
+    const { onDismiss, entityTypes } = this.props
+    const { uploading, file, newEntityType, isInvalidFile, isInvalidTSV } = this.state
+
+    const inputLabel = text => div({ style: { fontSize: 16, marginBottom: '0.3rem' } }, [text])
+
+    return h(Dropzone, {
```
Comment (katevoss) [inline src/libs/data-utils.js:239]: Is there a way to make the dropzone turn blue when a user is dragging a file over, like the other dropzones?

File: src/libs/data-utils.js [hunk 3]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+        if (definedTypeMatch) {
+          this.setState({ file, isInvalidFile: false, isInvalidTSV: false, newEntityType: definedTypeMatch[1] })
+        } else {
+          this.setState({ file: undefined, isInvalidTSV: true, isInvalidFile: false })
+        }
+      }
+    }, [
+      h(Modal, {
+        onDismiss,
+        title: 'Upload Table From .tsv File',
```
Comment (katevoss) [inline src/libs/data-utils.js:263]: I think we should make it clearer that the .tsv file is the way that a user adds data to their workspace. I suggest rewording the title:
> Upload data in a tsv file

possible secondary line below the title:
> To add data in a table to this workspace, upload a file with .tsv extension

File: src/libs/data-utils.js [hunk 4]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+          disabled: !file || uploading,
+          onClick: () => this.doUpload()
+        }, ['Upload'])
+      }, [
+        file && _.includes(_.toLower(newEntityType), entityTypes) && div({
+          style: { ...warningBoxStyle, marginBottom: '0.5rem', display: 'flex', alignItems: 'center' }
+        }, [
+          icon('warning-standard', { size: 24, className: 'is-solid', style: { flex: 'none', marginRight: '0.5rem', marginLeft: '-0.5rem' } }),
+          div([`Data with the type '${newEntityType}' already exists in this workspace. `,
+            'Uploading another load file for the same type may overwrite some entries.'])
```
Comment (katevoss) [inline src/libs/data-utils.js:274]: What is a load file? I know it's wording from FireCloud but I think we should clarify it here. 

Also, what do we recommend that a user does if `Data with the type ____ already exists in this workspace`? What safe action can they take?

File: src/libs/data-utils.js [hunk 5]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+        ]),
+        (isInvalidFile || isInvalidTSV) && div({
+          style: { color: colors.orange[0], fontWeight: 'bold', fontSize: 12, marginBottom: '0.5rem' }
+        }, [isInvalidFile ? 'Only .tsv files can be uploaded.' : 'File does not start with entity or membership definition.']),
```
Comment (katevoss) [inline src/libs/data-utils.js]: What is an `entity or membership definition`?

File: src/pages/workspaces/workspace/Data.js [hunk 6]
```diff
@@ -192,14 +192,21 @@ const WorkspaceData = _.flow(
...
 
     return div({ style: styles.tableContainer }, [
       !entityMetadata ? spinnerOverlay : h(Fragment, [
         div({ style: styles.dataTypeSelectionPanel }, [
-          div({ style: styles.dataTypeHeading }, 'Data Model'),
+          div({ style: { display: 'flex', alignItems: 'center', justifyContent: 'space-between' } }, [
+            div({ style: styles.dataTypeHeading }, 'Tables'),
+            linkButton({
+              disabled: !canEdit,
+              tooltip: canEdit ? 'Upload .tsv' : 'You do not have access add data to this workspace.',
```
Comment (katevoss) [inline src/pages/workspaces/workspace/Data.js:206]: What about changing this tooltip to
> Upload data

to focus on the outcome of the action, rather than the action itself?

File: src/pages/workspaces/workspace/Data.js [hunk 7]
```diff
@@ -511,7 +524,7 @@ const WorkspaceData = _.flow(
 
     return Utils.cond(
       [!amendedAttributes, () => undefined],
-      [_.isEmpty(amendedAttributes), () => 'No Local Variables defined'],
+      [_.isEmpty(amendedAttributes), () => 'No Workspace Data defined'],
```
Comment (katevoss) [inline src/pages/workspaces/workspace/Data.js:527]: How does a user add Workspace Data? This would be a good place to tell them how to add it. 
Also does a user define workspace data or add it? Verbage question.

File: src/libs/data-utils.js [hunk 8]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+        inputLabel('Selected File'),
```
Comment (katevoss) [inline src/libs/data-utils.js:279]: Could you swap the order to have the `Selected File` section appear after the `Drag or Click to select a .tsv file` box?

File: src/libs/data-utils.js [hunk 9]
```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
...
+        (file && file.name) || 'None',
```
Comment (katevoss) [inline src/libs/data-utils.js]: What about making the 'None' gray text, to make it clearer that it is secondary text?
Then once a user has uploaded a valid .tsv, make the filename color black and bold? Or something else to make it clear they were successful?

File: src/libs/data-utils.js (hunk 3 above)
Comment (zarsky-broad) [reply src/libs/data-utils.js:263]: Brad wanted "table" here, I'm gonna punt on wording stuff for now. We should def do a wording pass later. There's _a lot_ to clarify, and I think your suggestions are good.
